File: from-0201-to-0300/0238-product-of-array-except-self/solution.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        pre = [1] * len(nums)
        suf = [1] * len(nums)

        cur = 1
        for i, num in enumerate(nums):
            pre[i] = cur
            cur *= num

        cur = 1
        for i in range(len(nums)-1, -1, -1):
            suf[i] = cur
            cur *= nums[i]

        return [p * s for p, s in zip(pre, suf)]

File: from-0201-to-0300/0238-product-of-array-except-self/solution.go
func productExceptSelf(nums []int) []int {
    n := len(nums)
    pre := make([]int, n)
    suf := make([]int, n)

    currPre, currSuf := 1, 1
    for i, j := 0, n - 1; i < n; i, j = i + 1, j - 1 {
        pre[i] = currPre
        suf[j] = currSuf
        currPre *= nums[i]
        currSuf *= nums[j]
    }

    result := make([]int, n)
    for i := 0; i < n; i ++ {
        result[i] = pre[i] * suf[i]
    }

    return result
}